Zend/zend_compile.c

```diff
@@ -3334,8 +3402,20 @@ void zend_compile_return(zend_ast *ast) /* {{{ */
 		opline->op1.var = CG(context).fast_call_var;
 	}
 
-	if (CG(active_op_array)->fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
-		zend_emit_return_type_check(&expr_node, CG(active_op_array)->arg_info - 1);
+    // Generator return types are handled separately
+	if (!(CG(active_op_array)->fn_flags & ZEND_ACC_GENERATOR) && CG(active_op_array)->fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
```

Comment:
What about the case where `ZEND_ACC_GENERATOR` is not yet set - i.e. a return occurs before the first yield?

I expect it will still work (just with a redundant QM_ASSIGN opcode), but there should be a testcase for this.